File: android-sdk/modelhitch-core/src/main/kotlin/com/genoventureslabs/modelhitch/KeyStore.kt
package com.genoventureslabs.modelhitch

interface KeyStore {
    suspend fun get(providerId: String): String?
    suspend fun set(providerId: String, apiKey: String)
    suspend fun delete(providerId: String)
}

class MemoryKeyStore : KeyStore {
    private val keys = mutableMapOf<String, String>()

    override suspend fun get(providerId: String): String? = synchronized(keys) { keys[providerId] }

    override suspend fun set(providerId: String, apiKey: String) {
        synchronized(keys) { keys[providerId] = apiKey }
    }

    override suspend fun delete(providerId: String) {
        synchronized(keys) { keys.remove(providerId) }
    }
}

File: flutter-sdk/modelhitch_dart/lib/src/keystore.dart
abstract interface class KeyStore {
  Future<String?> get(String providerId);
  Future<void> set(String providerId, String apiKey);
  Future<void> delete(String providerId);
}

final class MemoryKeyStore implements KeyStore {
  final Map<String, String> _keys = {};

  @override
  Future<String?> get(String providerId) async => _keys[providerId];

  @override
  Future<void> set(String providerId, String apiKey) async {
    _keys[providerId] = apiKey;
  }

  @override
  Future<void> delete(String providerId) async {
    _keys.remove(providerId);
  }
}
